### PR TITLE
Refactor k-NN documentation

### PR DESCRIPTION
**Pull request by jmazanec15**
*Issue #, if available:*

*Description of changes:*
This PR refactors and extends the documentation for k-NN. I broke up the documentation into several different sections. I also moved a lot of content in the k-NN README and repo to here. In the future, I believe all usage information should be located in this repository. I will submit a PR to k-NN repo removing a lot of the usage documentation.

I marked in WIP because I will need painless extension documentation from @VijayanB. I will remove WIP status once I add that documentation.

I also believe I messed up some of the math formulas and other markdown elements. Please provide corrections where I may have gone wrong. Also, any other suggestions are appreciated as well.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/knn/knn-score-script.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+A space corresponds to the function used to measure the distance between 2 points in order to determine the k-nearest neighbors. From the k-NN perspective, a lower score equates to a closer and better result. This is the opposite of how Elasticsearch scores results, where a greater score equates to a better result. We include the conversion to Elasticsearch scores in the table below::
+
+spaceType | Distance Function | Elasticsearch Score
+:--- | :--- | :---
+l2 | \[ Distance(X, Y) = \sum_{i=1}^n (X_i - Y_i)^2 \] | 1 / (1 + Distance Function)
+cosinesimil | \[ Distance(X, Y) = {A &middot; B \over \|A\| &middot; \|B\|} \] | 1 + Distance Function)
+hammingbit | Distance = ones(X xor Y) | 1/(1 + Distance Function)
+
+
+Cosine similarity returns a number between -1 and 1, and because Elasticsearch relevance scores can't be below 0, the k-NN plugin adds 1 to get the final score.
```
Comment (keithhc2) [inline docs/knn/knn-score-script.md:186]: LaTeX's math formulas render better in HTML than in markdown, so we can use this block to show the formulas:

<table>
  <thead style="text-align: left">
  <tr>
    <th>spaceType</th>
    <th>Distance Function</th>
    <th>Elasticsearch Score</th>
  </tr>
  </thead>
  <tr>
    <td>l2</td>
    <td>\[ Distance(X, Y) = \sum_{i=1}^n (X_i - Y_i)^2 \]</td>
    <td>1 / (1 + Distance Function)</td>
  </tr>
  <tr>
    <td>cosinesimil</td>
    <td>\[ {A &middot; B \over \|A\| &middot; \|B\|} =

    {\sum_{i=1}^n (A_i &middot; B_i) \over \sqrt{\sum_{i=1}^n A_i^2} &middot; \sqrt{\sum_{i=1}^n B_i^2}}\]
    where \(\|A\|\) and \(\|B\|\) represent normalized vectors.</td>
    <td>1 + Distance Function</td>
  </tr>
  <tr>
    <td>hammingbit</td>
    <td style="text-align:center">Distance = countSetBits(X \(\oplus\) Y)</td>
    <td> 1 / (1 + Distance Function)</td>
  </tr>
</table>

Comment (jmazanec15) [reply docs/knn/knn-score-script.md:186]: Ah cool. I will update that. Thanks

File: docs/knn/index.md [hunk 2]
```diff
@@ -4,290 +4,14 @@ title: KNN
...
+This plugin supports three different methods for obtaining the k-nearest neighbors from an index of vectors. The first method takes an approximate nearest neighbor approach; it uses the HNSW algorithm to return the approximate k-nearest neighbors to a query vector. This algorithm sacrifices indexing speed and search accuracy in return for lower latency and more scalable search. To learn more about the algorithm, please refer to [nmslib's documentation](https://github.com/nmslib/nmslib/) or [the paper introducing the algorithm](https://arxiv.org/abs/1603.09320). The second method extends Elasticsearch's script scoring functionality to execute a brute force, exact k-NN search. With this approach, users are able to run k-NN search on a subset of vectors in their index (sometimes referred to as a pre-filter search). The third method adds the distance functions as painless extensions that can be used in more complex combinations.
```
Comment (vamshin) [inline docs/knn/index.md:15]: how about we classify approaches to 2 sections.

1. **Approximate k-NN**
        Description: blah.. blah..
        usecases: <couple sentences on when to apply this> 
        For more details refer https://github.com/jmazanec15/for-elasticsearch-docs/blob/knn-docs-refactor/docs/knn/approximate-knn.md
       
2. **Exact k-NN/Dynamic k-NN/custom scoring**
        Description: blah.. blah..
        usecases: <couple sentences on when to apply this> 
        For more details refer https://github.com/jmazanec15/for-elasticsearch-docs/blob/knn-docs-refactor/docs/knn/knn-score-script.md

3. **Exact k-NN/Dynamic k-NN/custom scoring with painless** 
    Description: This is extension to second method. In this approach the distance functions as painless extensions that can be used in more complex combinations. 
   usecases: <couple sentences on when to apply this> 
   For more details refer https://github.com/jmazanec15/for-elasticsearch-docs/blob/knn-docs-refactor/docs/knn/painless-functions.md

File: docs/knn/knn-score-script.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+## Getting started with the score script
+
+Similar to approximate nearest neighbor search, in order to use the score script on a body of vectors, you first must create an index with one or more `knn_vector` fields. If you intend to just use the script score approach (and not the approximate approach) `index.knn` can be set to `false` and `index.knn.space_type` does not need to be set. The space type can be chosen during search. See the [spaces section](#spaces) to see what spaces the knn score script suppports. Here is an example that creates an index with two `knn_vector` fields:
```
Comment (vamshin) [inline docs/knn/knn-score-script.md:15]: >> If you intend to just use the script score approach (and not the approximate approach) `index.knn` can be set to `false`

This should be put in **Bold** preferably kind of **Blurb** or may be just separate it as a new para and mention as Note. This is a very crucial point to pay attention.

File: docs/knn/index.md (hunk 2 above)
Comment (jmazanec15) [reply docs/knn/index.md:15]: Makes sense. I will update.

File: docs/knn/knn-score-script.md (hunk 3 above)
Comment (jmazanec15) [reply docs/knn/knn-score-script.md:15]: I also have as a tip here: https://github.com/opendistro/for-elasticsearch-docs/pull/396/files#diff-ba7ea95f84ff621a32b3ac90b917dc1956c4f8837b76f7106b4e9fb238fd1732R37

**Comment by jmazanec15** [comment]
I added painless documentation from #381 

@aetter, @keithhc2, @VijayanB, @vamshin could you all please review?

**Review by vamshin — APPROVED** [review]
LGTM! Thanks for refactoring docs and making more readable.

File: docs/knn/painless-functions.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+              "color": "BLUE"
+            }
+          }
+        }
+      },
+      "script": {
+        "source": "1.0 + cosineSimilarity(params.query_value, doc[params.field])", 
+        "params": {
+          "field": "my_vector",
+          "query_value": [9.9, 9.9],
```
Comment (VijayanB) [inline docs/knn/painless-functions.md:37]: can you remove "," at end

File: docs/knn/jni-library.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+title: JNI Library
+nav_order: 5
+parent: k-NN
+has_children: false
+---
+
+# JNI Library
+In order to integrate nmslib's approximate k-NN functionality, which is implemented in C++, into the k-NN plugin, which is implemented in Java, we created a Java Native Interface library. Check out [this wiki](https://en.wikipedia.org/wiki/Java_Native_Interface) to learn more about JNI. This library allows the k-NN plugin to leverage nmslib's functionality.
+
+## Artifacts
```
Comment (aetter) [inline docs/knn/jni-library.md:12]: I think this section is my only real beef with this PR. It feels very much like "if you're going to build this plugin yourself, here's some information you need to know," which is better-suited to the README. I think I'd prefer a single sentence in the previous paragraph that just says:

```
For more information about how we build the JNI library binary and how to get the most of it in your production environment, see [here](https://github.com/opendistro-for-elasticsearch/k-NN#jni-library-artifacts).
```

Comment (jmazanec15) [reply docs/knn/jni-library.md:12]: That makes sense. I will update

**Pull request merged**